cmd/influx_inspect/cardinality/cardinality.go

```diff
@@ -0,0 +1,174 @@
+	fs.StringVar(&cmd.dbPath, "db-path", "", "Path to database. Required.")
+	fs.IntVar(&cmd.concurrency, "c", 1, "Set worker concurrency. Defaults to one.")
+	fs.BoolVar(&cmd.detailed, "detailed", false, "Include counts for fields, tags, ")
+	fs.BoolVar(&cmd.exact, "exact", false, "Report exact counts")
+	fs.StringVar(&cmd.rollup, "rollup", "m", "Rollup level - t: total, d: database, r: retention policy, m: measurement")
+	fs.SetOutput(cmd.Stdout)
+	if err := fs.Parse(args); err != nil {
+		return err
+	}
+
+	if cmd.dbPath == "" {
+		return errors.New("path to database must be provided")
+	}
+
+	if _, ok := legalRollups[cmd.rollup]; !ok {
+		return fmt.Errorf("invalid rollup specified: %q", cmd.rollup)
+	}
+
+	factory := CreateNodeFactory(cmd.detailed, cmd.exact)
+
+	dbMap := factory.newNode(true)
+
+	g, ctx := errgroup.WithContext(context.Background())
+	g.SetLimit(cmd.concurrency)
+	processTSM := func(db, rp, id, path string) error {
+		file, err := os.OpenFile(path, os.O_RDONLY, 0600)
+		if err != nil {
+			_, _ = fmt.Fprintf(cmd.Stderr, "error: %s: %v. Skipping.\n", path, err)
+			return nil
+		}
+
+		reader, err := tsm1.NewTSMReader(file)
+		if err != nil {
+			_, _ = fmt.Fprintf(cmd.Stderr, "error: %s: %v. Skipping.\n", file.Name(), err)
+			// NewTSMReader won't close the file handle on failure, so do it here.
+			_ = file.Close()
+			return nil
+		}
+		defer func() {
+			// The TSMReader will close the underlying file handle here.
+			if err := reader.Close(); err != nil {
+				_, _ = fmt.Fprintf(cmd.Stderr, "error closing: %s: %v.\n", file.Name(), err)
+			}
+		}()
+
+		seriesCount := reader.KeyCount()
+		for i := 0; i < seriesCount; i++ {
+			func() {
+				key, _ := reader.KeyAt(i)
+				seriesKey, field, _ := bytes.Cut(key, []byte("#!~#"))
+				measurement, tags := models.ParseKey(seriesKey)
+				var totalDepth int
+				if cmd.rollup == "m" {
+					totalDepth = 3
+				} else if cmd.rollup == "r" {
+					totalDepth = 2
+				} else if cmd.rollup == "d" {
+					totalDepth = 1
+				} else {
+					totalDepth = 0
+				}
+				initRecord(dbMap, 0, totalDepth, db, rp, measurement, key, field, tags)
+			}()
+		}
+		return nil
+	}
+	done := ctx.Done()
+	err = reporthelper.WalkShardDirs(cmd.dbPath, func(db, rp, id, path string) error {
+		select {
+		case <-done:
+			return nil
+		default:
+			g.Go(func() error {
+				return processTSM(db, rp, id, path)
+			})
+			return nil
+		}
+	})
+
+	if err != nil {
+		_, _ = fmt.Fprintf(cmd.Stderr, "%s: %v\n", cmd.dbPath, err)
+		return err
+	}
+	err = g.Wait()
+	if err != nil {
+		_, _ = fmt.Fprintf(cmd.Stderr, "%s: %v\n", cmd.dbPath, err)
+		return err
+	}
+
+	tw := tabwriter.NewWriter(cmd.Stdout, 8, 2, 1, ' ', 0)
+
+	if err = factory.printHeader(tw); err != nil {
+		return err
+	}
+	if err = factory.printDivider(tw); err != nil {
+		return err
+	}
+	for d, db := range dbMap.children() {
+		for r, rp := range db.children() {
+			for m, measure := range rp.children() {
+				if cmd.rollup == "m" {
```

Comment:
Do we even need the check? I'd expect the children list to be empty here unless the rollup was for measurements

Reply:
Good catch. Leftover code from when I was summing up lower levels instead of maintaining separate counts at higher levels.